src/python/pants/backend/java/compile/javac.py

```diff
@@ -37,13 +53,71 @@ class CompiledClassfiles:
     digest: Digest
 
 
-@rule(level=LogLevel.DEBUG)
+class CompileResult(Enum):
+    SUCCEEDED = "succeeded"
+    FAILED = "failed"
+    DEPENDENCY_FAILED = "dependency failed"
+
+
+@dataclass(frozen=True)
+class FallibleCompiledClassfiles(EngineAwareReturnType):
+    description: str
+    result: CompileResult
+    output: CompiledClassfiles | None
+    exit_code: int
+    stdout: str | None = None
+    stderr: str | None = None
+
+    @classmethod
+    def from_fallible_process_result(
+        cls,
+        description: str,
+        process_result: FallibleProcessResult,
+        output: CompiledClassfiles | None,
+        *,
+        strip_chroot_path: bool = False,
+    ) -> FallibleCompiledClassfiles:
+        def prep_output(s: bytes) -> str:
+            return strip_v2_chroot_path(s) if strip_chroot_path else s.decode()
+
+        exit_code = process_result.exit_code
+        # TODO: Coursier renders this line on macOS.
+        stderr = "\n".join(
+            line
+            for line in prep_output(process_result.stderr).splitlines()
+            if line != "setrlimit to increase file descriptor limit failed, errno 22"
+        )
+        return cls(
+            description=description,
+            result=(CompileResult.SUCCEEDED if exit_code == 0 else CompileResult.FAILED),
+            output=output,
+            exit_code=exit_code,
+            stdout=prep_output(process_result.stdout),
+            stderr=stderr,
+        )
+
+    def level(self) -> LogLevel:
+        return LogLevel.ERROR if self.exit_code != 0 else LogLevel.INFO
+
+    def message(self) -> str:
```

Comment:
Known issue: until #12877 has landed, this will only render the first time it fails, and not if it hits cache.